sycl/doc/extensions/ExtendedAtomics/SYCL_INTEL_extended_atomics.asciidoc

```diff
@@ -113,389 +115,404 @@ In SYCL 1.2.1, support for floating-point types is limited to the +load+, +store
+  static constexpr memory_order default_write_order = memory_order_traits<DefaultOrder>::write_order;
+  static constexpr memory_order default_read_modify_write_order = DefaultOrder;
 
-|`operator+=(T operand)`
-|Equivalent to +fetch_add(operand)+.
+  bool is_lock_free() const noexcept;
 
-|+operator-=(T operand)+
-|Equivalent to +fetch_sub(operand)+.
+  explicit atomic_ref(T*&);
+  atomic_ref(const atomic_ref&) noexcept;
+  atomic_ref& operator=(const atomic_ref&) = delete;
 
-|`T operator++(int operand)`
-|Available only when: +T+ != +float+.  Equivalent to +fetch_add(operand)+.
+  void store(T* operand,
+    memory_order order = default_write_order) const noexcept;
 
-|+T operator--(int operand)+
-|Available only when: +T+ != +float+.  Equivalent to +fetch_sub(operand)+.
+  T* operator=(T* desired) const noexcept;
 
-|`T operator++()`
-|Available only when: +T+ != +float+.  Equivalent to `fetch_add(1) + 1`.
+  T* load(memory_order order = default_read_order) const noexcept;
 
-|+T operator--()+
-|Available only when: +T+ != +float+.  Equivalent to `fetch_sub(1) - 1`.
+  operator T*() const noexcept;
 
-|`T operator&=(T operand)`
-|Available only when: +T+ != +float+.  Equivalent to +fetch_and(operand)+.
+  T* exchange(T* operand,
+    memory_order order = default_read_modify_write_order) const noexcept;
 
-|`T operator\|= (T operand)`
-|Available only when: +T+ != +float+.  Equivalent to +fetch_or(operand)+.
+  bool compare_exchange_weak(T* &expected, T* desired,
+    memory_order success = default_read_modify_write_order,
+    memory_order failure = default_read_order) const noexcept;
 
-|`T operator^= (T operand)`
-|Available only when: +T+ != +float+.  Equivalent to +fetch_xor(operand)+.
+  bool compare_exchange_weak(T* &expected, T* desired,
+    memory_order order = default_read_modify_write_order) const noexcept;
 
-|
-|===
+  bool compare_exchange_strong(T* &expected, T* desired,
+    memory_order success = default_read_modify_write_order,
+    memory_order failure = default_read_order) const noexcept;
 
-==== Remove:
+  bool compare_exchange_strong(T* &expected, T* desired,
+    memory_order order = default_read_modify_write_order) const noexcept;
 
-"Available only when: +T+ != +float+" from definitions of +compare_exchange_strong+, +fetch_add+ and +fetch_sub+.
+  T* fetch_add(difference_type,
+    memory_order order = default_read_modify_write_order) const noexcept;
 
-=== Modify Section 5.2
+  T* fetch_sub(difference_type,
+    memory_order order = default_read_modify_write_order) const noexcept;
 
-==== From:
+  T* operator++(int) const noexcept;
+  T* operator--(int) const noexcept;
+  T* operator++() const noexcept;
+  T* operator--() const noexcept;
+  T* operator+=(difference_type) const noexcept;
+  T* operator-=(difference_type) const noexcept;
 
-The SYCL +atomic+ class can support 64 bit data types...
+};
 
-==== To:
+void atomic_fence(memory_order order):
```

Comment:
Am I right that the reason why `atomic_fence` is a free function and not a member of `item` or `nd_item` class (like `barrier`, for example) is that `atomic_fence` might be needed even in single-threaded application, just to ensure that instructions order will not be changed by compiler and/or HW implementation details - right?

Reply:
There are two additional reasons: 1) `atomic_thread_fence` is a free function in standard C++; and 2) atomic operations with `acquire` or `release` semantics behave very similarly to fences, and it seemed weird to limit one to ND-range kernels but not the other.